[quant] Implemented InputWeightObserver for Linear inputs

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59474 [quant] Input Weight Equalization - prepare modifications
* **#59090 [quant] Implemented InputWeightObserver for Linear inputs**

Summary: Implemented two observers (InputEqualObserver and WeightEqualObserver) which will be inserted into the graph during prepare_fx().

Test Plan: python test/test_quantization.py TestEqualizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28836954](https://our.internmc.facebook.com/intern/diff/D28836954)

angelayi committed:
[quant] Implemented InputWeightObserver for Linear inputs

Summary:

Test Plan: python test/test_quantization.py TestObservers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary:

Test Plan: python test/test_quantization.py TestObservers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary:

Test Plan: python test/test_quantization.py TestObservers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary:

Test Plan: python test/test_quantization.py TestObservers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on torch/quantization/observer.py:
```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
         return x_orig
 
 
+class InputWeightMinMaxObserver(PerChannelMinMaxObserver):
```
Let us prefix this with `_` since we don't want to expose this to the user. It might also be a good idea to put this into a separate file with the rest of the equalization stuff.

supriyar commented on torch/quantization/observer.py:
```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
+    def __init__(self, dtype=torch.quint8,
+                 qscheme=torch.per_channel_affine, reduce_range=False,
+                 quant_min=None, quant_max=None, factory_kwargs=None) -> None:
+        super(InputWeightMinMaxObserver, self).__init__(ch_axis=1, dtype=dtype,
```
Are we using the ch_axis field anywhere in the observer?

supriyar replied:
let us also have an input_dtype and weight_dtype to separate them out

supriyar commented on torch/quantization/observer.py:
```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
+                 qscheme=torch.per_channel_affine, reduce_range=False,
+                 quant_min=None, quant_max=None, factory_kwargs=None) -> None:
+        super(InputWeightMinMaxObserver, self).__init__(ch_axis=1, dtype=dtype,
+                                                        qscheme=qscheme,
```
I think qscheme is confusing here because we don't know if it corresponds to input or weight. Can we distinguish it? In the code we can assert that we only expect per-tensor for input.

supriyar commented on torch/quantization/observer.py:
```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
+
+        return x_orig, w_orig
+
+    def calculate_scale(self) -> torch.Tensor:
```
let's call this equalization_scale instead to not confuse with `scale` from qparams

supriyar commented on torch/quantization/observer.py:
```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
+        (scale_input, zero_point_input) = self._calculate_qparams(min_input_scaled, max_input_scaled)
+
+        # Calculate qparams for the weights
+        (scale_weight, zero_point_weight) = self._calculate_qparams(self.min_weights_row, self.max_weights_row)
```
we'd also need to multiply the inverse scale factors to the weights.

supriyar commented on torch/quantization/observer.py:
```diff
@@ -799,6 +802,129 @@ def forward(self, x_orig):
+        self.max_weights_col.copy_(max_weights_col)
+
+        # Calculate the min/max of weight rows
+        self.ch_axis = 0
```
rather than doing this, it might be a better idea to take an axis argument in _find_minmax

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary:

Test Plan: python test/test_quantization.py TestObservers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary:

Test Plan: python test/test_quantization.py TestObservers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary:

Test Plan: python test/test_quantization.py TestEqualizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28836954](https://our.internmc.facebook.com/intern/diff/D28836954)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -0,0 +1,173 @@
+
+    Given running min/max of the input columns as :math:`x_\text{min}` and :math:`x_\text{max}`,
+    and the running min/max of the weight columns as :math:`w_\text{min}` and :math:`w_\text{max}`,
+    scale :math:`S` is computed as:
```
nit: rename scale -> equalization_scale

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -0,0 +1,173 @@
+
+    The running minimum/maximum :math:`x_\text{min/max}` and
+    :math:`w_\text{min/max}` are computed in the same way as
+    :class:`~torch.quantization.observer.MinMaxObserver`, with the difference
```
Should this be PerChannelMinMaxObserver?

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -0,0 +1,173 @@
+    The scale factor :math:`S` is then computed as:
+
+    .. math::
+        S = \sqrt{\frac{x_{max} - x_{min}}{w_{max} - w{min}}}
```
Shouldn't the math be `\sqrt{\frac{w_{max} - w_{min}}{x_{max} - x{min}}}`

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -0,0 +1,173 @@
+    """
+
+    def __init__(self, input_dtype=torch.quint8, input_qscheme=torch.per_tensor_affine,
+                 input_quant_min=None, input_quant_max=None, weight_dtype=torch.quint8,
```
I think weight dtype should be torch.qint8

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -0,0 +1,173 @@
+        min_weights_ind = []
+        max_weights_ind = []
+        for i in range(num_row):
+            min_weights_ind.append(torch.where(w_orig[i] == self.weight_row_obs.min_vals[i])[0][0])
```
torch.where(cond) is equivalent to torch.nonzero(condition)
https://pytorch.org/docs/stable/generated/torch.nonzero.html#torch.nonzero

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -0,0 +1,173 @@
+            return torch.tensor([1.0]), torch.tensor([0]), torch.tensor([1.0]), torch.tensor([0])
+
+        (min_inputs, max_inputs) = self.get_input_minmax()
+        min_input_ind = torch.argmin(min_inputs)
+        max_input_ind = torch.argmax(max_inputs)
```
let's multiply the equalization_scale with the min_inputs/max_inputs and then compute the new min/max for the inputs.
So we can do torch.argmin(min_inputs * eq_scale) instead of torch.argmin(min_inputs) * eq_scale[idx]

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary:

Test Plan: python test/test_quantization.py TestEqualizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28836954](https://our.internmc.facebook.com/intern/diff/D28836954)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary:

Test Plan: python test/test_quantization.py TestEqualizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28836954](https://our.internmc.facebook.com/intern/diff/D28836954)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -0,0 +1,170 @@
+        min_weights_ind = []
+        max_weights_ind = []
+        for i in range(num_row):
+            min_weights_ind.append(torch.nonzero(w_orig[i] == self.weight_row_obs.min_vals[i])[0][0])
```
Do we still need to access [0][0] even after using `torch.nonzero`?

angelayi replied:
Yes, torch.nonzero seems to return an element in a 2D array.

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary:

Test Plan: python test/test_quantization.py TestEqualizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28836954](https://our.internmc.facebook.com/intern/diff/D28836954)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -0,0 +1,111 @@
+        weight_obs = _WeightEqualObserver(dtype=weight_qdtype,
+                                          qscheme=weight_qscheme)
+
+        width = np.random.randint(10) + 1
+        x_height = np.random.randint(10) + 2
+        w_height = np.random.randint(10) + 2
```
Why do we add the constant values here?

angelayi replied:
The constant values are to make sure that the width is greater than 0 and the heights are greater than 1. If the heights are equal to 1 then the min/max values will be the same and there will be a division by 0 error when calculating the scale.

angelayi replied:
I can also replace them with `w = np.random.randint(1, 10)` and `h = np.random.randint(2, 10)` if that's more readable.

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary:

Test Plan: python test/test_quantization.py TestEqualizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28836954](https://our.internmc.facebook.com/intern/diff/D28836954)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

jerryzh168 commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -0,0 +1,111 @@
+import torch
+from torch.testing._internal.common_quantization import QuantizationTestCase
+from torch.quantization.fx._equalize import (
+    _InputEqualObserver, _WeightEqualObserver, calculate_equalization_scale
```
nit: InputEqualizationObserver and WeightEqualizationObserver feels clearer..

jerryzh168 commented on torch/quantization/fx/_equalize.py:
```diff
@@ -0,0 +1,221 @@
+        return scale_weight, zero_point_weight
+
+
+def calculate_equalization_scale(input_obs: _InputEqualObserver, weight_obs: _WeightEqualObserver) -> torch.Tensor:
```
should we set_equalization_scale in convert instead? setting equalization scale in a `calculate_equalization_scale` doesn't sound very intuitive.

angelayi replied:
Sure, I will move it out and instead set the equalization scale in convert().

jerryzh168 commented on torch/quantization/fx/_equalize.py:
```diff
@@ -0,0 +1,221 @@
+        if qscheme not in {torch.per_tensor_affine, torch.per_tensor_symmetric}:
+            raise TypeError("Input qscheme must be per-tensor")
+
+        self.input_obs = PerChannelMinMaxObserver(ch_axis=1, dtype=dtype,
```
I think we should have an output_obs as well for output observer, right?

angelayi replied:
I thought the output observer was taken into account in the from_float() function and we would just replace the InputEqualizationObserver with a MinMaxObserver in prepare()?

supriyar replied:
The output observer is set to whatever is used for the input observer currently. So if the user wants to use something other than MinMax for the output, they wouldn't be able to.
We can add an optional argument to the InputEqualizationObserver that takes in output_observer and set it to default MinMaxObserver.

angelayi replied:
Oh ok, got it!

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary: Implemented two observers (InputEqualObserver and WeightEqualObserver) which will be inserted into the graph during prepare_fx().

Test Plan: python test/test_quantization.py TestEqualizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28836954](https://our.internmc.facebook.com/intern/diff/D28836954)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary: Implemented two observers (InputEqualObserver and WeightEqualObserver) which will be inserted into the graph during prepare_fx().

Test Plan: python test/test_quantization.py TestEqualizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28836954](https://our.internmc.facebook.com/intern/diff/D28836954)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary: Implemented two observers (InputEqualObserver and WeightEqualObserver) which will be inserted into the graph during prepare_fx().

Test Plan: python test/test_quantization.py TestEqualizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28836954](https://our.internmc.facebook.com/intern/diff/D28836954)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -0,0 +1,231 @@
+    def forward(self, x_orig):
+        # TODO: Allow for convoluational layers
+        if not (x_orig.ndim == 2):
+            raise ValueError("InputEqualObserver only supports Linear layers")
```
Update the message here to InputEqualizationObserver

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -0,0 +1,231 @@
+    def forward(self, w_orig):
+        # TODO: Allow for convoluational layers
+        if not (w_orig.ndim == 2):
+            raise ValueError("WeightEqualObserver only supports Linear layers")
```
same here too

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary: Implemented two observers (InputEqualObserver and WeightEqualObserver) which will be inserted into the graph during prepare_fx().

Test Plan: python test/test_quantization.py TestEqualizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28836954](https://our.internmc.facebook.com/intern/diff/D28836954)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

angelayi committed:
Update on "[quant] Implemented InputWeightObserver for Linear inputs"


Summary: Implemented two observers (InputEqualObserver and WeightEqualObserver) which will be inserted into the graph during prepare_fx().

Test Plan: python test/test_quantization.py TestEqualizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28836954](https://our.internmc.facebook.com/intern/diff/D28836954)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836954).

jerryzh168 reviewed (approved):
Looks good, thanks for addressing my comments!

jerryzh168 commented on torch/quantization/fx/_equalize.py:
```diff
@@ -0,0 +1,231 @@
+                                                  quant_max=quant_max,
+                                                  factory_kwargs=factory_kwargs)
+
+        if output_obs is None:
```
actually I think this should probably be a constructor as well? otherwise all the Tensors in the model will use the same output observer if user customizes it.

bigfootjon commented:
Closed by https://github.com/pytorch/pytorch/commit/cc03ea2c476b1137fb26f7fffac3b62dfcb6329c
